### PR TITLE
feat(forge): file cheatcodes

### PR DESCRIPTION
**Pull request by shekhirin**
## Motivation

https://github.com/foundry-rs/foundry/issues/1359 https://github.com/foundry-rs/foundry/issues/2006

## Solution

Introduce cheatcodes for common operations involving files: reading (fully or line by line), writing (fully or line by line), closing for reading and removing.

Thanks to [introduction](https://blog.soliditylang.org/2022/02/16/solidity-0.8.12-release-announcement/) of `string.concat` in Solidity 0.8.12, it's possible to easily write CSV rows to files with these cheatcodes by joining columns of data with delimiter. So I don't see clear need for CSV writing support.

I'm also open to comments especially regarding two things:
- It's possible to have only one instance of file open for reading, so you won't be able to have different offsets because files and offsets are linked by paths. The solution may be to link them by some kind of id (or file descriptor but that's not necessary imo).
- Semantics of `writeFile` and `writeLine` since they're behaving differently in terms of trailing newlines: former doesn't write it, and latter does.

### CODE REVIEW

File: testdata/cheats/File.t.sol [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        string memory path = "../testdata/fixtures/File/write_file.txt";
+        string memory data = "hello writable world";
+        cheats.writeFile(path, data);
+
+        assertEq(cheats.readFile(path), data);
+
+        cheats.removeFile(path);
+    }
+
+    function testWriteLine() public {
```
Comment (onbjerg) [inline testdata/cheats/File.t.sol:34]: How do these work in the real world? I.e. if I have two tests that write to the same file, will it work as expected?

I'm asking because one of primary use cases (IIRC) for writing to files was to record inputs and stuff in fuzz tests, so it's conceivable that this would be written to the same file in a format like `<test name>,<output>` or something.

**Review by onbjerg — COMMENTED** [review]
A small comment otherwise this looks really good :smile:

File: testdata/cheats/File.t.sol (hunk 1 above)
Comment (shekhirin) [reply testdata/cheats/File.t.sol:34]: it seems to me that it'll work the same way as writing to the same file from Rust? So if it's a POSIX-compliant system, most of the writes (especially small ones like lines) should be atomic and not interleavable, see for example https://stackoverflow.com/a/1154599

**Review by onbjerg — APPROVED** [review]
LGTM :)

File: evm/src/executor/inspector/cheatcodes/mod.rs [hunk 2]
```diff
@@ -85,6 +85,9 @@ pub struct Cheatcodes {
 
     /// Scripting based transactions
     pub broadcastable_transactions: VecDeque<TypedTransaction>,
+
+    //// Offsets for files opened for reading
+    pub file_reading_offsets: HashMap<String, usize>,
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/mod.rs:93]: would it make sense to keep the actual File buf here?

but I guess that's currently not possible due to how `Cheatcodes` inspector is created (Clone)

File: testdata/cheats/Cheats.sol [hunk 3]
```diff
@@ -103,4 +103,23 @@ interface Cheats {
     function startBroadcast(address) external;
     // Stops collecting onchain transactions
     function stopBroadcast() external;
+    // Reads the entire content of file to string, (path) => (data)
+    function readFile(string calldata) external returns (string memory);
+    // Reads next line of file to string, (path) => (line)
+    function readLine(string calldata) external returns (string memory);
```
Comment (mattsse) [inline testdata/cheats/Cheats.sol:109]: I've thought about how solidity `Fs` lib could look like:

https://github.com/mattsse/foundry/blob/e21fefd9488240e5f62378671be5da7691d6170b/evm-adapters/script-lib/Fs.sol

idea was to have a solidity struct that maps to a file and via using Interface you could call `std::fs::File` functions directly on it

having them all in the `VM` is fine though

File: testdata/cheats/File.t.sol (hunk 1 above)
Comment (mattsse) [reply testdata/cheats/File.t.sol:34]: yeh, also all fuzz cases for the _same_ test are called sequentially.
and even if there's another test that reads7writes the same file we can expect them to be atomic I think

File: evm/src/executor/inspector/cheatcodes/mod.rs (hunk 2 above)
Comment (shekhirin) [reply evm/src/executor/inspector/cheatcodes/mod.rs:93]: that's how I initially did it but yeah, due to `Clone` derive it's not possible unless we implement it on our own with some hashmap of bufs handling

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 4]
```diff
@@ -154,7 +161,70 @@ fn get_env(key: &str, r#type: ParamType, delim: Option<&str>) -> Result<Bytes, B
...
+    let mut reader = BufReader::new(file);
+    // Seek to read from offset
+    reader.seek(SeekFrom::Start(*offset as u64)).map_err(|err| err.to_string().encode())?;
+
+    let mut line: String = String::new();
+    let written = reader.read_line(&mut line).map_err(|err| err.to_string().encode())?;
+    // Remove one trailing newline character, preserving others for cases where it may be important
+    if line.ends_with(&['\r', '\n'][..]) {
+        line.pop();
+    }
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply evm/src/executor/inspector/cheatcodes/ext.rs]: yep fair, fixed

File: evm/src/executor/inspector/cheatcodes/mod.rs (hunk 2 above)
Comment (mattsse) [reply evm/src/executor/inspector/cheatcodes/mod.rs:93]: I faced similar issues in #2035

perhaps a better way would be to make two types

rn we have the `Inspectorstack` which is sort of like factory that can create new `Cheatcodes` inspectors, merely by cloning its `cheatcodes` field, because it can be initialized I think?

what if we replace clone with add dedicated `create_cheatcodes()` that clones the initialised subset and creates a new `Cheatcodes` object with an empty `Context` or something

I hope that made sense

```rust
impl Inspectorstack
{
  fn create_cheatcodes(&self) -> Cheatcodes {
    let subset = self.cheatcodes.clone();
    // merge all fields of subset to `Cheatcodes`
    Cheatcodes {... , context: Default::default()}
  }
}

```

Comment (onbjerg) [reply evm/src/executor/inspector/cheatcodes/mod.rs:93]: Yeah the reason we have `InspectorStack` cloneable is because we can configure the behavior of the inspectors (e.g. what inspectors are enabled/disabled). It also keeps an instance of `Cheatcodes` because some state needs to be persisted across some test boundaries, e.g. if you call `startPrank` in `setUp` this needs to apply for every test in that test contract.

Not sure if there's a better way to do that but I'd be keen to hear if there is because I don't really like how it works rn either.

File: testdata/cheats/File.t.sol (hunk 1 above)
Comment (mds1) [reply testdata/cheats/File.t.sol:34]: > I'm asking because one of primary use cases (IIRC) for writing to files was to record inputs and stuff in fuzz tests, so it's conceivable that this would be written to the same file in a format like `<test name>,<output>` or something.

Yep, you're thinking of #1022 (cc @transmissions11). Since runs of a fuzz test execute serially like @mattsse said, it seems we could call `writeLine` in the fuzz test to update the output file in each run, and that is arguably sufficient to close that issue. Two downsides though:
- You wouldn't be able to write a header row since we can't identify which fuzzer run we're in. Possible workaround is to add a cheatcode to return the current fuzz run number?
- Writing to the file in each fuzz run will be a lot slower than one batch write at the end.

We might also want to add a `vm.toString()` or `vm.stringify()` cheatcode which converts value types to a string. There are solidity libraries that help with this, but none are perfect (e.g. leading zeros are truncated if you stringify addresses, since they pass through uint)

Comment (onbjerg) [reply testdata/cheats/File.t.sol:34]: >You wouldn't be able to write a header row since we can't identify which fuzzer run we're in. Possible workaround is to add a cheatcode to return the current fuzz run number?

Arguably you could also write this header in `setUp` :)

Comment (mds1) [reply testdata/cheats/File.t.sol:34]: Ahh interesting, nice workaround! I don't love that it fragments file writing, e.g. some of the code for it is done in `setUp` and some in the test itself, but definitely ok for now

Comment (transmissions11) [reply testdata/cheats/File.t.sol:34]: > Writing to the file in each fuzz run will be a lot slower than one batch write at the end.

this is my primary concern

ffi is incredibly slow right now— hoping this much better

File: evm/src/executor/inspector/cheatcodes/mod.rs (hunk 2 above)
Comment (shekhirin) [reply evm/src/executor/inspector/cheatcodes/mod.rs:93]: > what if we replace clone with add dedicated create_cheatcodes() that clones the initialised subset and creates a new Cheatcodes object with an empty Context or something

yeah, that sounds good. Do I get it right that no other fields except `file_reading_offsets`/`opened_read_files` should be in that `Context` struct? I thought `corrected_nonce` might be a candidate but its resetting looks better in broadcast-related piece:
https://github.com/foundry-rs/foundry/blob/54f102bb601b5712fc7f480d5bf25ae851b745f6/evm/src/executor/mod.rs#L370-L384

Comment (shekhirin) [reply evm/src/executor/inspector/cheatcodes/mod.rs:93]: also, it changes behaviour of file reading a bit: previously I guess it was possible to start reading file in one test, and continue in another (given they're executed in order). Now, each test has its own set of files opened for reading and it's not shared with other tests.

File: testdata/cheats/File.t.sol (hunk 1 above)
Comment (shekhirin) [reply testdata/cheats/File.t.sol:34]: yeah it should be much faster

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";

contract ContractTest is Test {
    function setUp() public {}

    function _writeToFile(string memory path, string memory line) internal {
        string[] memory inputs = new string[](5);
        inputs[0] = "/Users/shekhirin/Projects/bench-forge-write/write-to-file.sh";
        inputs[1] = "-f";
        inputs[2] = path;
        inputs[3] = "-i";
        inputs[4] = line;

        vm.ffi(inputs);
    }

    function testFFI() public {
        for (uint i = 0; i < 1000; i++) {
            _writeToFile("ffi.txt", "gm");
        }

        vm.removeFile("ffi.txt");
    }

    function testWriteLine() public {
	for (uint i = 0; i < 1000; i++) {
            vm.writeLine("writeLine.txt", "gm");
	}

	vm.removeFile("writeLine.txt");
    }
}
```

```console
➜  bench-forge-write git:(main) ✗ ../foundry/target/release/forge test --ffi --match-test testFFI
[⠢] Compiling...
No files changed, compilation skipped

Running 1 test for test/Contract.t.sol:ContractTest
[PASS] testFFI() (gas: 4789029)
Test result: ok. 1 passed; 0 failed; finished in 1.81s

➜  bench-forge-write git:(main) ✗ ../foundry/target/release/forge test --match-test testWriteLine
[⠢] Compiling...
No files changed, compilation skipped

Running 1 test for test/Contract.t.sol:ContractTest
[PASS] testWriteLine() (gas: 650236)
Test result: ok. 1 passed; 0 failed; finished in 23.69ms
```

Comment (shekhirin) [reply testdata/cheats/File.t.sol:34]: yeah, looks like a general `std::process::Command` overhead
image

File: evm/src/executor/inspector/cheatcodes/mod.rs [hunk 5]
```diff
@@ -88,6 +92,22 @@ pub struct Cheatcodes {
...
+    //// Buffered readers for files opened for reading (path => BufReader mapping)
+    pub opened_read_files: HashMap<String, BufReader<File>>,
+}
+
+/// Every time we clone `Context`, we want it to be empty
+impl Clone for Context {
+    fn clone(&self) -> Self {
+        Default::default()
+    }
 }
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/mod.rs:111]: this is a suitable solution, because we will have a new,empty context for each test call.

I have some ideas on how to make this perhaps a bit more idiomatic in the multifork PR

**Review by mattsse — APPROVED** [review]
lgtm.

imo the names of the cheatcodes are ok, so they should be stable already, so that we can add to forge-std right away wdyt @gakonst ?

File: testdata/cheats/File.t.sol (hunk 1 above)
Comment (transmissions11) [reply testdata/cheats/File.t.sol:34]: awesome

**Comment by odyslam** [comment]
I completely missed this.. @shekhirin amazing job!

@mattsse @onbjerg what do you think of a vm.stringify() cheatcode? In my experience, stringifying was very not trivial in some cases.

**Comment by mds1** [comment]
+1 on a stringify cheatcode that has various overloads to convert value types + `bytes` to strings

**Comment by odyslam** [comment]
Attempting a quick implementation and will try to PR today.

The next step is to start thinking about how we can do read/write from JSON files

File: testdata/cheats/Cheats.sol (hunk 3 above)
Comment (gakonst) [reply testdata/cheats/Cheats.sol:109]: Feels like we can think of utilities built on top with forge std?
